### PR TITLE
[SHARE-465][Feature] Show full hierarchy in type facet

### PR DESCRIPTION
**Pull request by leb2dg**
** Depends on https://github.com/CenterForOpenScience/SHARE/pull/556

## Purpose

Show full hierarchy in type facet.

## Changes

screen shot 2017-01-13 at 12 01 19 pm

## Side effects

UI might look a little strange.

Small screen:
screen shot 2017-01-13 at 12 40 06 pm

Medium screen:
screen shot 2017-01-13 at 12 41 18 pm

Prevents:
screen shot 2017-01-13 at 12 09 06 pm

## Ticket

https://openscience.atlassian.net/browse/SHARE-465

### CODE REVIEW

File: app/components/search-facet-worktype-button/component.js [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
     selected: Ember.computed('selectedTypes.[]', function() {
         let selectedTypes = this.get('selectedTypes');
         return selectedTypes.contains(this.get('type'));
     }),
 
-    click() {
-        this.$().blur();
-        this.sendAction('onClick', this.get('type'));
+    label: Ember.computed('type', function() {
+        return this.get('type').replace(/^./, function(str) { return str.toUpperCase(); });
```
Comment (chrisseto) [inline app/components/search-facet-worktype-button/component.js]: http://emberjs.com/api/classes/Ember.String.html#method_capitalize

File: app/controllers/discover.js [hunk 2]
```diff
@@ -121,6 +127,39 @@ export default ApplicationController.extend({
         });
     },
 
+    transformTypes(obj) {
+        if (typeof (obj) !== 'object') {
+            return obj;
+        }
+        var keys = Object.keys(obj);
```
Comment (chrisseto) [inline app/controllers/discover.js]: Use [`for ... in`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) here

File: app/controllers/discover.js [hunk 3]
```diff
@@ -121,6 +127,39 @@ export default ApplicationController.extend({
...
+        var n = keys.length;
```
Comment (chrisseto) [inline app/controllers/discover.js]: [`let` is preferred to `var`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let)

`let` controls scope properly, `var` generally ends up being an unintended global variable.

File: app/controllers/discover.js [hunk 4]
```diff
@@ -121,6 +127,39 @@ export default ApplicationController.extend({
...
+        while (n--) {
+            var key = keys[n];
+            if (key === (lowKey = key.replace(/([A-Z])/g, ' $1').trim().toLowerCase())) {
+                obj[key] = this.transformTypes(obj[key]);
+                continue;
+            }
+            obj[lowKey] = this.transformTypes(obj[key]);
+            delete obj[key];
+        }
+        return (obj);
```
Comment (chrisseto) [inline app/controllers/discover.js]: No parens.

File: app/controllers/discover.js [hunk 5]
```diff
@@ -121,6 +127,39 @@ export default ApplicationController.extend({
...
+        if (typeof (obj) !== 'object') {
+            return obj;
+        }
+        var keys = Object.keys(obj);
+        var n = keys.length;
+
+        var lowKey;
+        while (n--) {
+            var key = keys[n];
+            if (key === (lowKey = key.replace(/([A-Z])/g, ' $1').trim().toLowerCase())) {
```
Comment (chrisseto) [inline app/controllers/discover.js]: You only need the `toLowerCase` here. The `replace` isn't actually doing anything

```javascript
'Foo Bar'.replace(/([A-Z])/g, ' $1').trim()
> "Foo  Bar"
```

File: app/controllers/discover.js [hunk 6]
```diff
@@ -121,6 +127,39 @@ export default ApplicationController.extend({
...
+                obj[key] = this.transformTypes(obj[key]);
+                continue;
+            }
+            obj[lowKey] = this.transformTypes(obj[key]);
```
Comment (chrisseto) [inline app/controllers/discover.js]: You only need to call `this.transformTypes` once here. (DRY it up)

File: app/controllers/discover.js [hunk 7]
```diff
@@ -121,6 +127,39 @@ export default ApplicationController.extend({
         });
     },
 
+    transformTypes(obj) {
```
Comment (chrisseto) [inline app/controllers/discover.js]: I'm not sure I see the point of this function, why wouldn't you just lowercase the keys whenever you use them?

File: app/controllers/discover.js [hunk 8]
```diff
@@ -121,6 +127,39 @@ export default ApplicationController.extend({
...
+                obj[key] = this.transformTypes(obj[key]);
+                continue;
+            }
+            obj[lowKey] = this.transformTypes(obj[key]);
+            delete obj[key];
+        }
+        return (obj);
+    },
+
+    getTypes() {
```
Comment (chrisseto) [inline app/controllers/discover.js]: Could you add an if guard here? If someone revisits this page without actually refreshing this request does not need to be made again.

File: app/controllers/discover.js [hunk 9]
```diff
@@ -49,6 +49,7 @@ export default ApplicationController.extend({
     numberOfResults: 0,
     took: 0,
     numberOfSources: 0,
+    types: '',
```
Comment (chrisseto) [inline app/controllers/discover.js]: Any reason that this defaults to a string? An empty object or null would be a bit more clear.
In general try to keep a single type per variable, especially when setting a default value.

File: app/components/search-facet-worktype-button/component.js (hunk 1 above)
Comment (leb2dg) [inline app/components/search-facet-worktype-button/component.js]: Oh snap

File: app/controllers/discover.js (hunk 9 above)
Comment (leb2dg) [inline app/controllers/discover.js]: Nope. Long story short, I changed it from an empty object to a string when debugging something and never changed it back.

File: app/controllers/discover.js (hunk 5 above)
Comment (leb2dg) [inline app/controllers/discover.js]: The model names are CamelCased: CreativeWork --> creative work

File: app/controllers/discover.js (hunk 7 above)
Comment (leb2dg) [inline app/controllers/discover.js]: The keys need to be lowercased and have a space between the words for the filters to work properly (at least as they are now). 

There is a lot of code that deals with the filters: 
* to and from url query params
* adding and removing from the facet
* pretty much every time a facet changes.

**Pull request merged**